### PR TITLE
improve perfomance of the domUtils.isShadowUI 

### PR DESCRIPTION
**Pull request by miherlosev**
@churkin, @AlexanderMoskovkin

Part number 2: improve perfomance of the domUtils.isShadowUI 

ShadowUI element is marked with the internal property `INTERNAL_PROPS.shadowUIElement` and `element.className` that contains `-hammerhead-shadow-ui` suffux.

ClassName uses for applying css styles and `INTERNAL_PROPS.shadowUIElement` uses for speed up the `IsShadowUIElement` check.

### CODE REVIEW

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
Optimization affects `getElementsBy*` methods. The average time of these methods on big read page (I saw an [HTMLEditor demos](https://demos.devexpress.com/ASPxHTMLEditorDemos/Default.aspx)) is about 5 %. It is difficult to see an optimization results.

This is why, I've checked performance on the synthetic page - [index.zip](https://github.com/DevExpress/testcafe-hammerhead/files/1032156/index.zip).

This page contains the following script code:
```js
<script>
			var elementCount = 1000;
			var testRoot     = document.getElementById('testRoot');
			
			for(var i = 0; i < elementCount; i++) {
				var divElement   = document.createElement('div');
				var inputElement = document.createElement('input');
				
				divElement.className = 'test';
				inputElement.name    = 'test';
				
				testRoot.appendChild(divElement);
				testRoot.appendChild(inputElement);
			}
			
			var countSearch = 100;
			
			for(var i = 0; i < countSearch; i ++) {
				document.getElementsByTagName('div');
				document.getElementsByClassName('test');
				document.getElementsByName('test');
				document.querySelectorAll('div');
			}
						
		</script>
```

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
Health Monitor run - https://testcafe-hhhm.devexpress.com/history?timestamp=1496054788471

**Comment by miherlosev** [comment]
Tests are green, health monitor is green.

@churkin @AlexanderMoskovkin  FPR

**Comment by miherlosev** [comment]
see detailed description - https://github.com/DevExpress/testcafe-hammerhead/pull/1150#issuecomment-304300552

File: src/client/sandbox/code-instrumentation/properties/index.js [hunk 1]
```diff
@@ -65,26 +64,17 @@ export default class PropertyAccessorsInstrumentation extends SandboxBase {
...
-
-                if (className.indexOf(SHADOW_UI_CLASSNAME.postfix) !== -1)
-                    elementCount++;
-            }
+            if (domUtils.isShadowUIElement(collection[i]))
+                shadowUIElementCount++;
         }
 
-        if (elementCount !== 0)
+        if (shadowUIElementCount !== 0)
```
Comment (churkin) [inline src/client/sandbox/code-instrumentation/properties/index.js]: ` if (shadowUIElementCount)`

File: src/client/sandbox/code-instrumentation/properties/index.js [hunk 2]
```diff
@@ -304,6 +294,18 @@ export default class PropertyAccessorsInstrumentation extends SandboxBase {
 
                     el.innerHTML = processedValue;
 
+                    if (domUtils.isBodyElement(el)) {
+                        var shadowUIRoot = this.shadowUI.select('.' + this.shadowUI.ROOT_CLASS, el)[0];
```
Comment (churkin) [inline src/client/sandbox/code-instrumentation/properties/index.js]: https://github.com/churkin/testcafe-hammerhead/blob/master/src/client/sandbox/shadow-ui.js#L238

File: src/client/sandbox/node/element.js [hunk 3]
```diff
@@ -573,6 +578,11 @@ export default class ElementSandbox extends SandboxBase {
 
         if (domUtils.isScriptElement(el))
             this.emit(this.SCRIPT_ELEMENT_ADDED, { el });
+
+        if (ElementSandbox._hasShadowUIParentOrContainsShadowUIClassPostfix(el)) {
+            ShadowUI.markElementAsShadow(el);
+            ShadowUI.markElementChildrenAsShadowRecursively(el);
```
Comment (churkin) [inline src/client/sandbox/node/element.js]: You can combine these two methods

File: src/client/sandbox/code-instrumentation/properties/index.js [hunk 4]
```diff
@@ -304,6 +294,18 @@ export default class PropertyAccessorsInstrumentation extends SandboxBase {
 
                     el.innerHTML = processedValue;
 
+                    if (domUtils.isBodyElement(el)) {
```
Comment (churkin) [inline src/client/sandbox/code-instrumentation/properties/index.js]: Why not use `onBodyContentChanged` https://github.com/churkin/testcafe-hammerhead/blob/master/src/client/sandbox/code-instrumentation/properties/index.js#L330 event?

Comment (miherlosev) [inline src/client/sandbox/code-instrumentation/properties/index.js]: `onBodyContentchanged` is called asynchronously.
There is test case where we need to sync flow - https://github.com/DevExpress/testcafe-hammerhead/pull/1150/files#diff-52837961726917bc92aefa76983a3eceR694

File: src/client/sandbox/code-instrumentation/properties/index.js (hunk 2 above)
Comment (miherlosev) [inline src/client/sandbox/code-instrumentation/properties/index.js]: This code will not create root if it doesn't exist.
The `shadowUI.getRoot` will create a root.
So, it is not a big problem and I will use `shadowUI.getRoot`.

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
FPR
ping @churkin

**Comment by miherlosev** [comment]
FPR
@churkin

**Pull request merged**